supervisor/api/host.py

```diff
@@ -240,8 +240,15 @@ async def advanced_logs_handler(
             try:
                 response = web.StreamResponse()
                 response.content_type = CONTENT_TYPE_TEXT
-                await response.prepare(request)
-                async for line in journal_logs_reader(resp, log_formatter):
+
+                async def finish_prepare(cursor: str):
+                    if cursor:
+                        response.headers["X-First-Cursor"] = cursor
+                    await response.prepare(request)
+
+                async for line in journal_logs_reader(
+                    resp, log_formatter, finish_prepare
+                ):
```

Comment:
It feels a bit weird that we hand-off control over when the `response.prepare` gets executed to the journal_logs_reader. Can the journal_logs_reader maybe return besides the generator something else? A bit funkty too but like so:

```python
generator, first_cursor = journal_logs_reader(resp, log_formatter)
if first_cursor:
    response.headers["X-First-Cursor"] = cursor
    await response.prepare(request)
async for line in generator:
    ...
```

:thinking:

Reply:
I rewrote it to return the cursor in a tuple on every iteration, it's closest to what you propose. To get even closer, you'll have to perform the first anext on the generator to get the first cursor, so it'll end up like this:

```python
generator = journal_logs_reader(resp, log_formatter)
first_cursor, first_line = await anext(generator)

async def wrapped_generator(first):
    yield first
    async for _, x in generator:
        yield x

if first_cursor:
    response.headers["X-First-Cursor"] = first_cursor
await response.prepare(request)

async for line in wrapped_generator(first_line):
    await response.write(line.encode("utf-8") + b"\n")
```

This is maybe slightly better than the conditional return in the loop but let me know what you think.

Reply:
Cool, yeah I think that rewrite is better.

> This is maybe slightly better than the conditional return in the loop but let me know what you think.

anext and generator is quite a bit less readable, let's keep the condition.